### ISSUE TITLE
discrepency between CFE and analyzer's operator == override

### ISSUE THREAD

**Issue by srawlins** [dart-lang/sdk#50628]
See this example from @yanok ([dartpad](https://dartpad.dev/?id=b38530d5832a0daa809f346508ca8b48)):

```dart
abstract class A {
  @override
  bool operator ==(dynamic other);
}

class B implements A {}

void test(dynamic x, dynamic y, dynamic z) {
  print(x == y || x == z);
}

void main() {
  var b1 = B(), b2 = B();
  B? b3;
  print(b1 == null);
  print(b1 == b2);
  print(b3 == b2);
  print(b3 == null);
  test(b1, b2, b3);
}
```

This program runs without error, but analyzer reports an `invalid_implementation_override ` error at line 6 (`class B implements A {}`):

> 'Object.==' ('bool Function(Object)') isn't a valid concrete implementation of 'A.==' ('bool Function(dynamic)').

There is either a bug in analyzer or CFE.

cc @scheglov

**Comment by scheglov** [comment]
The analyzer implements what is written in the specification.
<img width="609" alt="image" src="https://user-images.githubusercontent.com/384794/205739688-a4e8a5fa-c5af-433a-bff4-123db701ea58.png">
Specifically "`F` must then be a subtype of `F′`."
And `bool Function(Object)` is not a subtype of `bool Function(dynamic)` because `dynamic` is not a subtype of `Object`.
It accepts for example `null`, which is never passed to `operator==`, but the specification does not make an exception.
@eernstg

**Comment by johnniwinther** [comment]
Yes. The ability to allow the legacy handling of the `==` argument, required a lax handling of this case. As far as I remember, because we can't (couldn't?) tell whether the `dynamic` parameter was user-provided or from untyped legacy code.

**Comment by eernstg** [comment]
It sounds like we can decide that the current behaviors of the CFE and the analyzer are OK. There is a slight inconsistency, but it will go away when the support for legacy code is removed. Moreover, there is no soundness issue as far as I can see:

As specified, operator `==` is never invoked with the argument null (even `super ==` is subject to the special null checks before the method is called), so there's no way an invocation of an operator `==` (like the one in `Object`, inherited by `B`), requiring an argument of type `Object`, can be invoked with the argument null. So it doesn't matter that the type `B` seems to promise that we can pass null as the argument. Also, there is no way to tear off operator `==`.

**Comment by eernstg** [comment]
@sigmundch, I believe the web compilers may treat operator `==` differently: They do not generate code to perform the null checks at the call site; instead, they generate code at the beginning of the body of operator `==` to check for null and return true/false as appropriate. Is that correctly described? Do you have any worries about keeping the current CFE behavior (where there is no compile-time error in the situation where a `==` signature with type `bool Function(dynamic)` is implemented by an implementation with type `bool Function(Object)`)?
